### PR TITLE
Add API and UI to remove requirements installed by Conda

### PR DESCRIPTION
**Pull request by mvdbeek**
This is currently limited to the Conda dependency resolver, but other
resolver would only need to implement the resolver.uninstall method and
set can_uninstall_dependencies to True.

To use this endpoint, DELETE to
`<galaxy_url>/tools/<tool_id>/dependencies`.

In this PR I have also added the possibility to POST to <galaxy_url>/tools/<tool_id>/dependencies, which will install the dependencies, in favor of the older `<galaxy_url>/tools/<tool_id>/install_dependencies` endpoint which we will deprecate (when ?).

If this is accepted, I will enhance the new manage tool dependencies page so that we can remove and reinstall dependencies from there.

### CODE REVIEW

**Comment by jmchilton** [comment]
@mvdbeek Are you willing to work on these endpoint URLs a bit?

I'd like to maintain

```
POST /api/tools/{tool_id}/install_dependencies
```

for backward compat. but I guess going forward I'd like to see the following for install and uninstall:

```
POST /api/tools/{tool_id}/dependencies
DELETE /api/tools/{tool_id}/dependencies
```

It feels more RESTful to me.

**Comment by mvdbeek** [comment]
Oh, sure, that makes sense!

**Comment by jmchilton** [comment]
Did you mean to add #3729 into this PR?

**Comment by mvdbeek** [comment]
damnit, too much switching branches :(

**Comment by mvdbeek** [comment]
Alright, this should be OK now.

**Review by martenson — APPROVED** [review]
looks great

**Comment by martenson** [comment]
should we add `POST /api/tools/{tool_id}/install_dependencies` to https://github.com/galaxyproject/galaxy/issues/3280 ?

**Comment by jmchilton** [comment]
@martenson I'd say -0 on deprecation - the amount of technical debt we reduce is not worth the effort and the cost of not maintaining backward compat.

Update: I mean I'm fine if we all agree it is deprecated - I just don't think it is worth removing per se.

**Comment by mvdbeek** [comment]
@galaxybot test this

**Comment by martenson** [comment]
@jmchilton not worth removing - worth announcing though?

**Comment by mvdbeek** [comment]
I have fixed a small problem in 6902e62 and I have added a button to actually uninstall the conda env in the UI in 95106ea.

**Comment by martenson** [comment]
@galaxybot test this

**Comment by martenson** [comment]
@galaxybot test this

**Comment by martenson** [comment]
@galaxybot test this

**Comment by martenson** [comment]
@galaxybot test this

**Comment by martenson** [comment]
@galaxybot test this

**Comment by martenson** [comment]
@galaxybot test this

**Comment by martenson** [comment]
Thanks @mvdbeek !

File: lib/galaxy/tools/deps/resolvers/conda.py [hunk 1]
```diff
@@ -116,6 +117,25 @@ def get_option(name):
...
+            all_resolved = [self.resolve(requirement) for requirement in requirements]
+            all_resolved = [r for r in all_resolved if r.dependency_type]
+        if not all_resolved:
+            return None
+        environments = set([os.path.basename(dependency.environment_path) for dependency in all_resolved])
+        return_codes = [self.conda_context.exec_remove([env]) for env in environments]
+        final_return_code = 0
+        for env, return_code in zip(environments, return_codes):
+            if return_code == 0:
+                log.debug("Conda environment '%s' sucessfully removed." % env)
```
Comment (nsoranzo) [inline lib/galaxy/tools/deps/resolvers/conda.py:133]: s/sucessfully/successfully/

File: lib/galaxy/tools/deps/resolvers/conda.py [hunk 2]
```diff
@@ -116,6 +117,25 @@ def get_option(name):
     def clean(self, **kwds):
         return self.conda_context.exec_clean()
 
+    def uninstall(self, requirements):
+        """Uninstall requirements installed by install_all or multiple install statements."""
+        all_resolved = [r for r in self.resolve_all(requirements) if r.dependency_type]
```
Comment (nsoranzo) [inline lib/galaxy/tools/deps/resolvers/conda.py:122]: This will break when `self.resolve_all()` returns `False` in line 163.

**Pull request merged**